src/operator/filter_map.rs

```diff
@@ -1,145 +1,484 @@
+        F: Fn(Self::ItemRef<'_>) -> bool + 'static,
     {
-        self.filter_map_keys_generic(func)
+        self.circuit()
+            .add_unary_operator(FilterKeys::new(filter_func), self)
     }
 
-    /// Both filter and map keys in the input batch with a user-provided
-    /// closure.
-    ///
-    /// For each `(key, value)` pair in the input batch, the operator computes
-    /// `func(key)`. If the result is `None`, the pair is dropped.  If the
-    /// result is `Some(newkey)`, the `(newkey, value)` pair is added to the
-    /// output batch.
-    ///
-    /// This method returns output batches of type [`OrdIndexedZSet`].
-    /// Use [`Self::filter_map_keys_generic`] to create other types of output
-    /// batches.
-    pub fn filter_map_keys<F, T>(
-        &self,
-        func: F,
-    ) -> Stream<Circuit<P>, OrdIndexedZSet<T, CI::Val, CI::R>>
+    fn map_generic<F, T, O>(&self, map_func: F) -> Stream<Circuit<P>, O>
     where
-        CI: BatchReader<Time = ()> + 'static,
-        CI::Val: Clone + Ord,
-        T: Ord + Clone + 'static,
-        F: Fn(&CI::Key) -> Option<T> + Clone + 'static,
+        F: Fn(Self::ItemRef<'_>) -> T + Clone + 'static,
+        O: Batch<Key = T, Val = (), Time = (), R = Self::R> + Clone + 'static,
     {
-        self.filter_map_keys_generic(func)
+        self.circuit().add_unary_operator(
+            MapKeys::new(map_func.clone(), move |x| (map_func)(&x)),
+            self,
+        )
     }
 
-    /// Both filter and map keys in the input batch with a user-provided
-    /// closure.
-    ///
-    /// For each `(key, value)` pair in the input batch, the operator computes
-    /// `func(key)`. If the result is `None`, the pair is dropped.  If the
-    /// result is `Some(newkey)`, the `(newkey, value)` pair is added to the
-    /// output batch.
-    ///
-    /// This method is generic over output batch type.
-    /// Use `Self::filter_map` to return a batch of type [`OrdZSet`].
-    /// Use `Self::filter_map_keys` to return a batch of type
-    /// [`OrdIndexedZSet`].
-    pub fn filter_map_keys_generic<CO, F>(&self, func: F) -> Stream<Circuit<P>, CO>
+    fn map_index_generic<F, KT, VT, O>(&self, map_func: F) -> Stream<Circuit<P>, O>
+    where
+        F: Fn(Self::ItemRef<'_>) -> (KT, VT) + 'static,
+        O: Batch<Key = KT, Val = VT, Time = (), R = Self::R> + Clone + 'static,
+    {
+        self.circuit().add_unary_operator(
+            Map::new(move |kv: (Self::ItemRef<'_>, &())| map_func(kv.0)),
+            self,
+        )
+    }
+
+    fn flat_map_generic<F, I, O>(&self, func: F) -> Stream<Circuit<P>, O>
+    where
+        F: Fn(Self::ItemRef<'_>) -> I + 'static,
+        I: IntoIterator + 'static,
+        O: Batch<Key = I::Item, Val = (), Time = (), R = Self::R> + Clone + 'static,
+    {
+        self.circuit().add_unary_operator(
+            FlatMap::new(move |kv: (Self::ItemRef<'_>, &())| {
+                func(kv.0).into_iter().map(|x| (x, ()))
```

Comment:
isn't there a cost to be paid to assemble these tuples with an empty component?

Reply:
Hopefully not. Can't say I 100% trust the optimizer though.

Reply:
It's a zero sized type, it doesn't exist at runtime